packages/augur-core/source/contracts/trading/ZeroXTrade.sol

```diff
@@ -367,16 +451,67 @@ contract ZeroXTrade is Initializable, IZeroXTrade, IERC1155 {
         }
     }
 
-    /// @dev Decode ERC-1155 asset data from the format described in the AssetProxy contract specification.
+    /// @dev Decode MultiAsset asset data from the format described in the AssetProxy contract specification.
     /// @param _assetData AssetProxy-compliant asset data describing an ERC-1155 set of assets.
-    /// @return The ERC-1155 AssetProxy identifier, the address of the ERC-1155
+    /// @return The ERC-1155 AssetProxy identifier, the address of this ERC-1155
     /// contract hosting the assets, an array of the identifiers of the
     /// assets to be traded, an array of asset amounts to be traded, and
     /// callback data.  Each element of the arrays corresponds to the
     /// same-indexed element of the other array.  Return values specified as
     /// `memory` are returned as pointers to locations within the memory of
     /// the input parameter `assetData`.
     function decodeAssetData(bytes memory _assetData)
+        public
+        view
+        returns (
+            bytes4 _assetProxyId,
+            address _tokenAddress,
+            uint256[] memory _tokenIds,
+            uint256[] memory _tokenValues,
+            bytes memory _callbackData,
+            address _kycToken
+        )
+    {
+         // Read the bytes4 from array memory
+        assembly {
+            _assetProxyId := mload(add(_assetData, 32))
+            // Solidity does not require us to clean the trailing bytes. We do it anyway
+            _assetProxyId := and(_assetProxyId, 0xFFFFFFFF00000000000000000000000000000000000000000000000000000000)
+        }
+
+        require(_assetProxyId == MULTI_ASSET_PROXY_ID, "WRONG_PROXY_ID");
+
+        uint256[] memory _amounts;
+        bytes[] memory _nestedAssetData;
+
+        // Slice the selector off the asset data
+        bytes memory _noSelectorAssetData = _assetData.slice(4, _assetData.length);
+
+        (_amounts, _nestedAssetData) = abi.decode(_noSelectorAssetData, (uint256[], bytes[]));
+        
+        // Validate storage refs against the decoded values.
+        {
+            require(_amounts.length == 3);
+            require(_amounts[0] == 1);
+            require(_amounts[1] == 0);
+            require(_amounts[2] == 0);
+            require(_nestedAssetData[1].equals(cashAssetData));
+            require(_nestedAssetData[2].equals(shareAssetData));
+        }
```

Comment:
For posterity: we don't need to check the length of nestedAssetData == 3, because `_amounts` is what drives the actual multi-asset process.

Reply:
Also protected from mismatch inside multi-asset proxy

```
                // Revert if number of elements in `amounts` differs from number of elements in `nestedAssetData`
                if sub(amountsLen, nestedAssetDataLen) {
                    // Revert with `Error("LENGTH_MISMATCH")`
```